website/docs/language/expressions/custom-validation-rules.mdx

```diff
@@ -0,0 +1,316 @@
+---
+page_title: Custom Validation Rules - Configuration Language
+description: >-
+  Validate requirements for variables, outputs, and within lifecycle blocks so Terraform can produce better error messages in context.
+---
+
+# Custom Validation Rules
+
+You can create validation checks with custom error messages for several types of objects in a configuration. Custom validations are optional, but they can help capture assumptions that might be only implied, helping future maintainers understand the configuration design and intent. They also return useful information about errors earlier and in context, helping consumers more easily diagnose issues in their configurations.
+
+You can create custom validations with the following types of expressions.
+- Add [`validation` blocks](input-variable-validation) inside input `variable` blocks.
+- Add [`precondition` and `postcondition`](#preconditions-and-postconditions) blocks inside `lifecycle` blocks and `precondition` blocks inside `output blocks`.
+
+All of these methods require a [`condition` attribute](#condition-expressions) that describes the validation requirements and an [`error_message` attribute](#error-messages) that contains explanatory text that Terraform will display to the user.
+
+-> **Note:** Input variable validation is available in Terraform CLI v0.13.0 and later. Preconditions and postconditions are available in Terraform CLI v1.2.0 and later.
+
+## When Terraform Evaluates Custom Validations
+
+Terraform will evaluate the conditions specified in `validation`,`precondition`, and `postcondition` blocks as early as possible.
```

Comment:
We might want to call out here that variable validations are always evaluated immediately, as they can only refer to the variable value itself.

Reply:
thank you! added this in and will push it up!